src/Spanner/Database.php

```diff
@@ -0,0 +1,1313 @@
+     * @codingStandardsIgnoreEnd
+     *
+     * @param string $table The table to mutate.
+     * @param KeySet $keySet The KeySet to identify rows to delete.
+     * @param array $options [optional] Configuration options.
+     * @return Timestamp The commit Timestamp.
+     */
+    public function delete($table, KeySet $keySet, array $options = [])
+    {
+        $mutations = [$this->operation->deleteMutation($table, $keySet)];
+
+        $session = $this->selectSession(SessionPoolInterface::CONTEXT_READWRITE);
+
+        $options['singleUseTransaction'] = $this->configureTransactionOptions();
+
+        try {
+            return $this->operation->commit($session, $mutations, $options);
+        } finally {
+            $session->setExpiration();
+        }
+    }
+
+    /**
+     * Run a query.
+     *
+     * Example:
+     * ```
+     * $result = $database->execute('SELECT * FROM Posts WHERE ID = @postId', [
+     *     'parameters' => [
+     *         'postId' => 1337
+     *     ]
+     * ]);
+     *
+     * $firstRow = $result
+     *     ->rows()
+     *     ->current();
+     * ```
+     *
+     * ```
+     * // Execute a read and return a new Snapshot for further reads.
+     * $result = $database->execute('SELECT * FROM Posts WHERE ID = @postId', [
+     *      'parameters' => [
+     *         'postId' => 1337
+     *     ],
+     *     'begin' => true
+     * ]);
+     *
+     * $result->rows()->current();
+     *
+     * $snapshot = $result->snapshot();
+     * ```
+     *
+     * ```
+     * // Execute a read and return a new Transaction for further reads and writes.
+     * $result = $database->execute('SELECT * FROM Posts WHERE ID = @postId', [
+     *      'parameters' => [
+     *         'postId' => 1337
+     *     ],
+     *     'begin' => true,
+     *     'transactionType' => SessionPoolInterface::CONTEXT_READWRITE
+     * ]);
+     *
+     * $result->rows()->current();
+     *
+     * $transaction = $result->transaction();
+     * ```
+     *
+     * @codingStandardsIgnoreStart
+     * @see https://cloud.google.com/spanner/reference/rpc/google.spanner.v1#google.spanner.v1.ExecuteSqlRequest ExecuteSqlRequest
+     * @codingStandardsIgnoreEnd
+     *
+     * @codingStandardsIgnoreStart
+     * @param string $sql The query string to execute.
+     * @param array $options [optional] {
+     *     Configuration Options.
+     *
+     *     See [TransactionOptions](https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.TransactionOptions)
+     *     for detailed description of available transaction options.
+     *
+     *     Please note that only one of `$strong`, `$minReadTimestamp`,
+     *     `$maxStaleness`, `$readTimestamp` or `$exactStaleness` may be set in
+     *     a request.
+     *
+     *     @type array $parameters A key/value array of Query Parameters, where
```

Comment:
Because we provide value types for values which could be ambiguous, we feel confident that we are able to accurately determine types for any valid value. That said, users could provide the params and paramTypes directly, bypassing the mapping entirely. This is not currently documented, as you can see, though we can certainly add it.